feat: Improved Linked Wearables documentation (#638)
decentraland/documentation: _posts/wearables/2022-04-05-linked-wearables.md

![]({{ site.baseurl }}/images/media/linkedw-proposal.png)

After the DAO has approved the proposal, a set of Third Party managers will be set to operate with the all of the Linked Wearables features.

## Creating a Linked Wearable Collection

> Linked wearables are grouped in collections that can be created, edited and deleted by Third Party Managers. Each collection can contain an arbitrary number of Linked Wearables.

Author: In this case, the comma is not required as it's an enumeration of things that can be done to the collections.

To create a new Linked Wearable follow these steps:

1. Create a new Linked Wearables Collection.

   ![]({{ site.baseurl }}/images/media/linkedw-collection.png)

2. Choose a name for the collection and an ID.

   ![]({{ site.baseurl }}/images/media/linkedw-name-collection.png)

> We strongly recommend that you choose a name that is unique and descriptive for the collection and **as the ID, the address of the NFT contract** that will be mapped to the Linked Wearables. Using the NFT address as the ID of you collection will prove to be helpful when creating the API.

Author: I've changed a bit the emphasis.

### Editing the collection name

To edit the name of a Linked Wearable Collection follow these steps:

1. Move the mouse over the collection name and click on the red pencil icon that appears on hover.

   ![]({{ site.baseurl }}/images/media/linked-wearables/edit-collection-name.png)

2. Choose a new name for the collection and click on the save button.

   ![]({{ site.baseurl }}/images/media/linked-wearables/edit-collection-name-modal.png)

### Editing the collection ID or URN

To edit the ID or URN of a Linked Wearable Collection follow these steps:

1. Click on the meatballs menu (three horizontal dots) on the far right of the set of buttons. A dropdown will appear. Select _Edit URN_.

   ![]({{ site.baseurl }}/images/media/linked-wearables/edit-collection-urn.png)

2. Write the new ID or URN for the collection and click _Save_.

   ![]({{ site.baseurl }}/images/media/linked-wearables/edit-collection-urn-modal.png)

The ID or URN of the collection **MUST be unique**, changing it to an existent one will fail.

### Deleting the collection

> To delete Linked Wearable Collection follow these steps:

Reviewer: On both delete and update collection id or urn I would repeat the message about that it can be done only if there are no published wearables.

Useful because when reading the docs the person might look for these titles and ignore the message above.

is created, you'll be able to upload your wearables' 3D models to the Builder.

## Uploading models one by one

It's possible to, as it already happens with standard wearables, upload your wearables' 3D models one by one.

To do so, follow these steps:

1. Click on the _New items_ button.

   ![]({{ site.baseurl }}/images/media/linked-wearables/new-items-collection.png)

2. Select the _Singe items_ option.

   ![]({{ site.baseurl }}/images/media/linked-wearables/multiple-items-upload.png)

3. Follow the steps to upload and configure your wearable as it is described in the _Uploading an item_ section in the [wearables editor user guide]({{ site.baseurl }}{% post_url /wearables/2021-05-31-wearables-editor-user-guide %}).

## Uploading models in bulk

As Linked Wearable collections can contain a big number of items, it is possible to upload the 3D models and the information of the wearables in bulk.

The process of uploading wearables in bulk involves uploading a ZIP file for each wearable containing the following:

> - The 3D model files of the wearable (GLB, GLTFs, texture files, etc).

Author: I've added "required" at the start of each of the bullet points, making emphasis on its requirability, so it's clear which of the is mandatory or not.

> - A `wearable.json` file containing the information of the wearable.

Reviewer 2: Can add a link to the `wearable.json` below?

> - A `wearable.json` file containing the information of the wearable.

Author: I've added "required" at the start of each of the bullet points, making emphasis on its requirability, so it's clear which of the is mandatory or not.

- All the information about the wearable categories and which to choose can be found in the [creating wearables guidelines]({{ site.baseurl }}{% post_url /wearables/2021-05-31-creating-wearables %}).
- The `id` field is optional and can be used to:
  - Create a wearable with an specific ID.
  - Update a wearable in Bulk (which will be seen in the `Making changes in bulk` section)
- The `id` field must contain the whole ID of the wearable, that is, the third party record id, the collection id and the item id `urn:decentraland:matic:collections-thirdparty:third-party-id:collection-id:item-id`. We recommend the id to be formed as `urn:decentraland:matic:collections-thirdparty:third-party-id:contract-address:token-id`, that is, if our wearable will be mapped to the token `0` in the NFT contract `0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d`, we recommend the URN to be `urn:decentraland:matic:collections-thirdparty:third-party-id:0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d:0`.
- The representations array will contain the information about how each body shape will look like. Each wearable MUST contain at least one representation (it can have one or the two of them), that is, taking

>   into consideration the body shapes that we currently have, either 'urn:decentraland:off-chain:base-avatars:BaseMale' or 'urn:decentraland:off-chain:base-avatars:BaseFemale'. Each representation will describe which models will be used for each body shape.

Reviewer 2: ```suggestion
  into consideration the body shapes that we currently have, either `urn:decentraland:off-chain:base-avatars:BaseMale` or `urn:decentraland:off-chain:base-avatars:BaseFemale`. Each representation will describe which models will be used for each body shape.
```

8. Success! Your items are now available in your collection.

   ![]({{ site.baseurl }}/images/media/linked-wearables/edit-multiple-files-success.png)

## Seeing the wearable in world

Linked Wearables can be seen in world to review how the model will work once published.

To be able to see a wearable in world, follow these steps:

1. Click on the meatballs menu (three horizontal dots) on the right of the item that you want to see in world. A dropdown will appear. Select _See in world_.

   ![]({{ site.baseurl }}/images/media/linked-wearables/see-in-world-button.png)

2. The Decentraland World will open. Navigate to your backpack to see the wearable.

   ![]({{ site.baseurl }}/images/media/linked-wearables/see-in-world-item.png)

## Publishing wearables

Once your wearables are ready, they must be published for curation. Publishing wearables **will consume slots**, a slot per wearable will be consumed upon publishing.

To publish your wearables, you need to:

1. Select the items to be published and wait for the Curators Committee to approve them

   ![]({{ site.baseurl }}/images/media/linkedw-publish.png)

> Note that it does not matter in which network your NFT project is: Ethereum, Polygon, Binance, etc. The API is needed to map the owner of the NFT to the 3D model (wearable) submitted to Decentraland. Lets imagine that you have an NFT project with only 10 tokens minted. Therefore you have created and submitted 10 different 3D models (wearables) to Decentraland. Now, you need to create an API that Decentraland can use to know which wearable must be displayed to each token holder of your project. To do that, you will need to have in hand the tokens' owner and the `urn` of each item submitted to Decentraland. The item's urn can be found on each item detail page:

Author: I've moved it and re-written it.

7. Be patient, this might take a while!

   ![]({{ site.baseurl }}/images/media/linked-wearables/edit-multiple-files-uploading.png)

8. Success! Your items are now available in your collection.

   ![]({{ site.baseurl }}/images/media/linked-wearables/edit-multiple-files-success.png)

# Submitting your Linked Wearables

Your Linked Wearables need to go through a publishing and curation process as the regular wearables do. Although the publication and curation process is not the same
as the one for the regular wearables, it keeps the same steps, all items must be first be published to later be curated by an assigned curator.

The following sections will show you how to publish your Linked Wearables to be curated.

## Publishing wearables

Once your wearables are ready, they must be published for curation. Your wearables are published in groups of items, you can choose which items are ready to be curated
by selecting them and clicking the `Publish` button. A slot per wearables will be consumed upon publishing and the third party managers won't be able to publish any
more wearables until the ones that are already published are curated.

To publish your wearables, you need to:

> 1. Select the items to be published and wait for the Curators Committee to approve them

Reviewer: I the quetion "Why should I select the items to be published intead of publishing them all?" worth answering here?

Author: That's a great question. I think that if we wanted, we could explain a lot of the decisions taken, but I think that it will extend a lot the documentation.

Reviewer: Fair enough. Let's keep it simple

## API

In order for Linked Wearables to work, the third parties need to provide an API that will be queried by the Decentraland services (Catalyst) with 2 endpoints:

1. `@GET /registry/:registry-id/address/:address/assets` - Retrieves a list of assets associated with a given address
2. `@GET /registry/:registry-id/address/:address/assets/:id` - Validates if a DCL asset is owned by a user.

**Technical details and examples [here](https://github.com/decentraland/adr/blob/main/docs/ADR-42-third-party-assets-integration.md#third-party-resolver).**

### Endpoint `@GET /registry/:registry-id/address/:address/assets`

If you've followed the steps to create a new collection and the steps to either edit the item's id or URN or to create items in bulk, you already have:

- Created a collection which its ID or URN is the contract address of the NFTs that will represent or map to the Linked Wearables.

> - Created wearables which IDs or URNs are the token ids of the NFTs that will represent or map to the Linked Wearables.

Reviewer 2: ```suggestion
- Created wearables whose IDs or URNs are the token ids of the NFTs that will represent or map to the Linked Wearables.
```

```

The API can do the following:

1. Query the contract `0xc04528c14c8ffd84c7c1fb6719b4a89853035cdd` with the user's address by using `balanceOf(0x0f5d2fb29fb7d3cfee444a200298f468908cc942)`, retrieving the amount of
   tokens owned by the user, in this case, 2.
2. Query the contract `0xc04528c14c8ffd84c7c1fb6719b4a89853035cdd` two times `tokenOfOwnerByIndex(0x0f5d2fb29fb7d3cfee444a200298f468908cc942, 0)`, `tokenOfOwnerByIndex(0x0f5d2fb29fb7d3cfee444a200298f468908cc942, 1)`
   returning the token ids 58 and 59.
3. Respond with:

```json
{
  "address": "0x0f5d2fb29fb7d3cfee444a200298f468908cc942",
  "assets": [
    {
      "id": "0xc04528c14c8ffd84c7c1fb6719b4a89853035cdd:58",
      "amount": 1,
      "urn": {
        "decentraland": "urn:decentraland:matic:collections-thirdparty:cryptohats:0xc04528c14c8ffd84c7c1fb6719b4a89853035cdd:58"
      }
    },
    {
      "id": "0xc04528c14c8ffd84c7c1fb6719b4a89853035cdd:59",
      "amount": 1,
      "urn": {
        "decentraland": "urn:decentraland:matic:collections-thirdparty:cryptohats:0xc04528c14c8ffd84c7c1fb6719b4a89853035cdd:59"
      }
    }
  ],

>   "total": 100,

Reviewer 2: ```suggestion
  "total": 2,
```

address of the owner of the token and, if it matches, respond with the required information about the asset.

As an example, by querying the API to get the information about the asset with id `0xc04528c14c8ffd84c7c1fb6719b4a89853035cdd:1` owned by the `0x0f5d2fb29fb7d3cfee444a200298f468908cc942` address:

```bash
curl https://api.cryptohats.io/registry/cryptohats/address/0x0f5d2fb29fb7d3cfee444a200298f468908cc942/assets/0xc04528c14c8ffd84c7c1fb6719b4a89853035cdd:1
```

The API can do the following:

1. Query the contract `0xc04528c14c8ffd84c7c1fb6719b4a89853035cdd` retrieved from the query by splitting the asset id by `:`, with the user's address also retrieved from the query, by using
   the method `ownerOf(1)` that will return the address `0x0f5d2fb29fb7d3cfee444a200298f468908cc942` if the asset is owned by the user.
2. Respond with:

```json
{
  "id": "0xc04528c14c8ffd84c7c1fb6719b4a89853035cdd:1",
  "amount": 1,
  "urn": {
    "decentraland": "urn:decentraland:matic:collections-thirdparty:cryptohats:0xc04528c14c8ffd84c7c1fb6719b4a89853035cdd:1"
  }
}

> ```

Reviewer 2: I would clarify that this is JUST an example of the API that will work ONLY if you have an NFT project and you will have one 3d model for each token minted. Also, that each project may use its own logic. The important thing is to map token ownership with the wearables submitted to DCL.

Author: I've added the `Important notice` down below with that information.

```

The API can do the following:

1. Query the contract `0xc04528c14c8ffd84c7c1fb6719b4a89853035cdd` retrieved from the query by splitting the asset id by `:`, with the user's address also retrieved from the query, by using
   the method `ownerOf(1)` that will return the address `0x0f5d2fb29fb7d3cfee444a200298f468908cc942` if the asset is owned by the user.
2. Respond with:

```json
{
  "id": "0xc04528c14c8ffd84c7c1fb6719b4a89853035cdd:1",
  "amount": 1,
  "urn": {
    "decentraland": "urn:decentraland:matic:collections-thirdparty:cryptohats:0xc04528c14c8ffd84c7c1fb6719b4a89853035cdd:1"
  }
}
```

## Important notice and considerations

The examples provided above on how to build the API are just examples, it's not required to follow them. The implementations of the API can vary completely, for example, the information can be retrieved from different sources (an indexer like [TheGraph](https://thegraph.com/), a Database previously populated or multiple indexers or contracts) and a NFT can be mapped to multiple decentraland wearables instead of having a 1:1 mapping with the NFTs.

> Note that the URNs in the examples use `matic` for the protocol. This means that the third party registry smart contract has been deployed in the Matic network **BUT** it is not related to or limiting where your project can be created. Every urn for linked wearables items will use `matic` as the protocol. Remember that linked wearables work for any NFT project created at any network (Ethereum, Polygon, Binance, etc). So, if your project has been created in Ethereum, the urn of the item will have `matic` in the protocol and it is ok.

Reviewer: "the urn of the item will have `matic` in the protocol and it is ok."

Would it be nice to explain that it says matic because the information of the third party is stored in the matic network?

Reviewer: @nachomazzara already mentioned this on another comment 👍

Author: This was added in the important notice section!

The purpose of the API is to map an owner of an NFT with a wearable (or multiple wearables) submitted to Decentraland and this mapping is done through the ID or URN of the wearable which can be found on each item detail page:

![]({{ site.baseurl }}/images/media/linked-wearables/item-urn.png)

# Closing up

> ## Summary

Reviewer: Would be nice that each bullet of the summery has an anchor to the respective documentation in this document. At least the main ones.

For example: 'Create a DAO proposal in the category “Linked Wearables Registry”. You’ll need to:' could have a link to 'http://0.0.0.0:4000/decentraland/linked-wearables/#dao-proposal' (with the proper domain)

Author: Changed the summary to have links in the bullet points that had one!

Reviewer: 👏

2. Wait one week for DAO's approval. **Ask your community for help!** they can vote for you to be accepted. Remember that MANA, Names, Lands, and/or Estates are needed to vote in the DAO.
3. **Submit your linked wearables** in the [Builder](https://builder.decentraland.org/) using your available Slots.

   1. Create a third party collection. We recommend using as its ID or URN the contract address of the NFTs that will represent or be mapped to the Linked Wearables.

>    2. Create or upload your wearables either by uploading them one by one or in bulk.

Reviewer 2: I would re-add this somewhere

```
Note that the URN uses `matic` for the protocol. This means that the third party registry smart contract has been deployed in the Matic network **BUT** it is not related to or limiting where your project can be created. Every urn for linked wearables items will use `matic` as the protocol. Remember that linked wearables work for any NFT project created at any network. So, if your project has been created in Ethereum, the urn of the item will have `matic` in the protocol and it is ok.
```

Author: Added this information to the `important notice` in the `Building the API` section.
